File: components/tic-tac-toe/tictactoe.ts
// simple implementation of tic tac toe copy 

type Player = 'X' | 'O';
type Cell = Player | null;
export type State = Cell[];  // mutable array for easy indexing

export const X: Player = 'X';
export const O: Player = 'O';
export const EMPTY: null = null;
const WIN_COMBOS: readonly [number, number, number][] = [
  [0, 1, 2], [3, 4, 5], [6, 7, 8],
  [0, 3, 6], [1, 4, 7], [2, 5, 8],
  [0, 4, 8], [2, 4, 6],
];

export const initialState = (): State => Array(9).fill(EMPTY);

export const player = (state: State): Player => {
  const xs = state.filter(c => c === X).length;
  const os = state.filter(c => c === O).length;
  return xs <= os ? X : O;
};

export const actions = (state: State): number[] =>
  state
    .map((c, i) => (c === EMPTY ? i : -1))
    .filter(i => i >= 0);

export const result = (state: State, i: number): State => {
  if (state[i] !== EMPTY) throw new Error('Invalid move');
  const next = state.slice();
  next[i] = player(state);
  return next;
};

export const winner = (state: State): Player | null => {
  for (const [a, b, c] of WIN_COMBOS) {
    const v = state[a];
    if (v && v === state[b] && v === state[c]) return v;
  }
  return null;
};

export const terminal = (state: State): boolean =>
  Boolean(winner(state)) || state.every(c => c !== EMPTY);

export const utility = (state: State): number => {
  const w = winner(state);
  return w === X ? 1 : w === O ? -1 : 0;
};

// Simple memoization cache for minimax
const cache = new Map<string, [number, number | null]>();
const keyOf = (state: State) => state.map(c => c ?? '-').join('');

const minimaxValue = (state: State): [number, number | null] => {
  const key = keyOf(state);
  if (cache.has(key)) return cache.get(key)!;

  if (terminal(state)) {
    const util = utility(state);
    const res: [number, null] = [util, null];
    cache.set(key, res);
    return res;
  }

  const turn = player(state);
  let bestVal = turn === X ? -Infinity : Infinity;
  let bestAct: number | null = null;

  for (const a of actions(state)) {
    const [val] = minimaxValue(result(state, a));
    if ((turn === X && val > bestVal) || (turn === O && val < bestVal)) {
      bestVal = val;
      bestAct = a;
    }
  }

  const res: [number, number | null] = [bestVal, bestAct];
  cache.set(key, res);
  return res;
};

export const minimax = (state: State): number | null => minimaxValue(state)[1];

/**
 * Apply a move for current player, or return null if invalid or game over.
 */
export function applyMove(state: State, i: number): State | null {
  if (state[i] !== EMPTY || winner(state) !== null) {
    return null;
  }
  return result(state, i);
}

File: backend/games/tic-tac-toe/tic-tac-toe.py
"""
Tic Tac Toe Player
"""

X, O, EMPTY = "X", "O", None


def initial_state():
    """
    Returns starting state of the board.
    """

    return [[EMPTY] * 3 for _ in range(3)]


def player(board):
    """
    Returns player who has the next turn on a board.
    """

    xs = sum(row.count(X) for row in board)
    os = sum(row.count(O) for row in board)
    return X if xs <= os else O


def actions(board):
    """
    Returns set of all possible actions (i, j) available on the board.
    """

    return {
        (i, j)
        for i, row in enumerate(board)
        for j, cell in enumerate(row)
        if cell is EMPTY
    }


def result(board, action):
    """
    Returns the board that results from making move (i, j) on the board.
    """
    i, j = action
    if board[i][j] is not EMPTY:
        raise ValueError("Invalid action")
    new = [row[:] for row in board]
    new[i][j] = player(board)
    return new


def winner(board):
    """
    Returns the winner of the game, if there is one.
    """

    lines = (
        board[:]  # rows
        + [[board[r][c] for r in range(3)] for c in range(3)]  # cols
        + [
            [board[i][i] for i in range(3)],
            [board[i][2 - i] for i in range(3)],
        ]  # diags
    )
    for line in lines:
        if line[0] is not EMPTY and line.count(line[0]) == 3:
            return line[0]
    return None


def terminal(board):
    """
    Returns True if game is over, False otherwise.
    """

    return winner(board) is not None or all(cell is not EMPTY for row in board for cell in row)


def utility(board):
    """
    Returns 1 if X has won the game, -1 if O has won, 0 otherwise.
    """

    win = winner(board)
    return 1 if win == X else -1 if win == O else 0


def minimax(board):
    """
    Returns the optimal action for the current player on the board.
    """
    
    if terminal(board):
        return None

    def max_value(b, α, β):
        if terminal(b):
            return utility(b), None
        v, mv = float("-inf"), None
        for a in actions(b):
            val, _ = min_value(result(b, a), α, β)
            if val > v:
                v, mv = val, a
                α = max(α, v)
            if α >= β:
                break
        return v, mv

    def min_value(b, α, β):
        if terminal(b):
            return utility(b), None
        v, mv = float("inf"), None
        for a in actions(b):
            val, _ = max_value(result(b, a), α, β)
            if val < v:
                v, mv = val, a
                β = min(β, v)
            if α >= β:
                break
        return v, mv

    _, move = (
        
        max_value(board, float("-inf"), float("inf"))
        if player(board) == X
        else min_value(board, float("-inf"), float("inf"))
    )
    return move
